apps/opencs/view/render/brushdraw.cpp

```diff
@@ -0,0 +1,308 @@
+    end.z() = std::numeric_limits<int>::lowest();
+    osg::Vec3d direction = end - start;
+
+    // Get intersection
+    osg::ref_ptr<osgUtil::LineSegmentIntersector> intersector (new osgUtil::LineSegmentIntersector(
+        osgUtil::Intersector::MODEL, start, end) );
+    intersector->setIntersectionLimit(osgUtil::LineSegmentIntersector::NO_LIMIT);
+    osgUtil::IntersectionVisitor visitor(intersector);
+
+    visitor.setTraversalMask(SceneUtil::Mask_Terrain);
+
+    mParentNode->accept(visitor);
+
+    for (osgUtil::LineSegmentIntersector::Intersections::iterator it = intersector->getIntersections().begin();
+         it != intersector->getIntersections().end(); ++it)
+    {
+        osgUtil::LineSegmentIntersector::Intersection intersection = *it;
+
+        // reject back-facing polygons
+        if (direction * intersection.getWorldIntersectNormal() > 0)
+        {
+            continue;
+        }
+
+        return intersection.getWorldIntersectPoint().z();
+    }
+    return 0.0f;
+}
+
+void CSVRender::BrushDraw::buildPointGeometry(const osg::Vec3d& point)
+{
+    osg::ref_ptr<osg::Geometry> geom (new osg::Geometry());
+    osg::ref_ptr<osg::Vec3Array> vertices (new osg::Vec3Array());
+    osg::ref_ptr<osg::Vec4Array> colors (new osg::Vec4Array());
+    const float brushOutlineHeight (1.0f);
+    const float crossHeadSize (8.0f);
+    osg::Vec4f lineColor(1.0f, 1.0f, 1.0f, 0.6f);
+
+    vertices->push_back(osg::Vec3d(
+        point.x() - crossHeadSize,
+        point.y() - crossHeadSize,
+        getIntersectionHeight(osg::Vec3d(
+            point.x() - crossHeadSize,
+            point.y() - crossHeadSize,
+            point.z()) ) + brushOutlineHeight));
+    colors->push_back(lineColor);
+    vertices->push_back(osg::Vec3d(
+        point.x() + crossHeadSize,
+        point.y() + crossHeadSize,
+        getIntersectionHeight(osg::Vec3d(
+            point.x() + crossHeadSize,
+            point.y() + crossHeadSize,
+            point.z()) ) + brushOutlineHeight));
+    colors->push_back(lineColor);
+    vertices->push_back(osg::Vec3d(
+        point.x() + crossHeadSize,
+        point.y() - crossHeadSize,
+        getIntersectionHeight(osg::Vec3d(
+            point.x() + crossHeadSize,
+            point.y() - crossHeadSize,
+            point.z()) ) + brushOutlineHeight));
+    colors->push_back(lineColor);
+    vertices->push_back(osg::Vec3d(
+        point.x() - crossHeadSize,
+        point.y() + crossHeadSize,
+        getIntersectionHeight(osg::Vec3d(
+            point.x() - crossHeadSize,
+            point.y() + crossHeadSize,
+            point.z()) ) + brushOutlineHeight));
+    colors->push_back(lineColor);
+
+    geom->setVertexArray(vertices);
+    geom->setColorArray(colors, osg::Array::BIND_PER_VERTEX);
+    geom->addPrimitiveSet(new osg::DrawArrays(osg::PrimitiveSet::LINES, 0, 4));
+    mGeometry = geom;
+}
+
+void CSVRender::BrushDraw::buildSquareGeometry(const float& radius, const osg::Vec3d& point)
```

Comment:
Just for future reference: there is no point to pass float by references, it does not make code faster.